### ISSUE TITLE
HTTP status is being overwritten in route file.

### ISSUE THREAD

**Issue by ofuochi** [lukeautry/tsoa#513]
## Sorting

* **I'm submitting a ...**
  - [X] bug report
  - [ ] feature request
  - [ ] support request

* I confirm that I
  - [X] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Expected Behavior

I was expecting that the error status code  be retained instead of overriding 

## Current Behavior

The auto-generated `route.ts` file overwrites the error status in the local `fail` method of the `authenticateMiddleware` function as shown. This changes the intended HTTP error that's been returned to the client even though it might have been set to something different than 401

```ts
...
const fail = function(error: any) {
 responded++;
 if (responded == security.length && !success) {
   error.status = 401;  // why this?
   next(error)
 }
}
...
```

## Possible Solution

A possible solution is to check that `error.status` is falsy before overwriting it. Like so...

```ts
...
const fail = function(error: any) {
 responded++;
 if (responded == security.length && !success) {
   error.status = (error && error.status) || 401;  //  this is better IMO
   next(error)
 }
}
...
```

## Steps to Reproduce

1. Write a custom error class which extends from `Error`
2. Declare a `status` which is an HTTP status code
3. Set the error status to 409 (forbidden) for example
4. Throw the error in the `expressAuthentication` method

## Context (Environment)
Version of the library: 2.5.6
Version of NodeJS: 6.11.3
* Confirm you were using yarn not npm: [x]

## Detailed Description

Below is my custom error class 

```ts
export class HttpError extends Error {
    readonly status: number;

    constructor(status?: number, message?: string, name?: string) {
        super(message);
        Error.captureStackTrace(this, this.constructor);

        this.status = status || 500;
        this.name = name || this.name;
        this.message =
            message ||
            http.STATUS_CODES[this.status] ||
            "An error occurred during the request.";
    }
}
```

I've decorated my controller action like so
```ts
...
@Post()
    @Security("X-Auth-Token", ["admin"])
    @Response(httpStatus.FORBIDDEN, http.STATUS_CODES[httpStatus.FORBIDDEN])
    public async create(@Body() input: CreateTenantInput) {
        await this.checkBadRequest(plainToClass(CreateTenantInput, input));
        await this.checkConflict(await this._tenantService.get(input.name));
        return this._tenantService.create(input.name, input.description);
    }
...
```

Somewhere in the `expressAuthentication` function, I throw this error like so...
```ts
...
    if (!isIdValid(tenantId))
        throw new HttpError(
            httpStatus.FORBIDDEN, // I'm setting 409 here and not 401
            `${tenantId} is not a valid ${X_TENANT_ID} header`
        );
...
```
As you can see, 409 should be the HTTP status and should not be overwritten to 401.

**Comment by dgreene1** [comment]
@ofuochi I agree that this should be fixed. Since you basically found the fix already, can you please submit a PR?

Please add this change to all 4 of the templates (express, koa, hapi, and custom):

`error.status = (error && error.status) || 401;`

**Comment by ofuochi** [comment]
I'll gladly do that.

**Comment by dgreene1** [comment]
Thanks @ofuochi. I published your PR in v2.5.7. Welcome to the tsoa contributor club! :)

**Comment by ofuochi** [comment]
I'll be happy to do more

**Comment by dgreene1** [comment]
@ofuochi wonderful. Here’s the list of issues that we’ve identified as being high value: https://github.com/lukeautry/tsoa/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22
